# Arreglos

Los arreglos, y matrices, son tipos de datos estructurados homogéneos (un solo tipo de dato), optimizados para el cálculo usando vectores y matrices. Los arrays suelen contener números enteros, reales y complejos. 

Python no tiene soporte integrado para arreglos. Para esto, utilizamos el librería [NumPy](https://numpy.org/), cuya documentación se encuentra en [NumPy documentation](https://numpy.org/doc/stable/).

<div align="center">
    <img src="https://github.com/MCD-ESPOL/PCD/blob/main/notebooks/images/logo.svg?raw=true" width="15%">
</div>

## Importar librerías externas

In [ ]:
# Importe la librería numpy, con: import numpy as np
# Importe la libreria matplotlib, con: import matplotlib.pyplot as plt


## De lista a arreglos

Para crear [un arreglo a partir de una lista](https://numpy.org/doc/stable/user/basics.creation.html#converting-python-sequences-to-numpy-arrays), utilizaremos la función **array**.



<div align="center">
    <img src="https://github.com/MCD-ESPOL/PCD/blob/main/notebooks/images/arr_ini.png?raw=true" width="40%">
</div>

In [ ]:
# Convierta las listas en los arreglos nombres_librosArr, genero_literarioArr y ventasArr

nombres_libros = ['El señor de los anillos', 'Cien años de soledad', 'Harry Potter', '1984', 'Orgullo y prejuicio', 'Crimen y castigo', 'El gran Gatsby', 'Matar un ruiseñor', 'La Odisea', 'Don Quijote de la Mancha']
genero_literario = ['Fantasía', 'Realismo mágico', 'Fantasía', 'Distopía', 'Novela romántica', 'Novela psicológica', 'Ficción moderna', 'Novela clásica', 'Epopeya', 'Novela clásica']
ventas = [150, 120, 200, 90, 80, 110, 95, 130, 75, 160]




In [ ]:
# Muestre los arreglos nombres_librosArr, genero_literarioArr y ventasArr


## Operaciones con arreglos

Numpy ofrece un grupo de [funciones agregadas](https://www.tutorialgateway.org/python-numpy-aggregate-functions/), como: **sum**, **mean**, **std**,  **max**, **min**, **argmin**, **argmax** y muchas más.

La lista completa de [funciones matemáticas](https://numpy.org/doc/stable/reference/routines.math.html) puede encontrarla en la documentación de Numpy.




<div align="center">
    <img src="https://github.com/MCD-ESPOL/PCD/blob/main/notebooks/images/arr_aggr.png?raw=true" width="60%">
</div>



In [ ]:
# Muestre el valor máximo del arreglo ventasArr


In [ ]:
# Obtenga el índice del valor máximo del arreglo ventasArr


# Con el valor anterior, indexe arreglos nombres_librosArr y genero_literarioArr
# Muestre los elementos obtenidos de la indexación




In [ ]:
# Obtenga y muestre el total de ventasArr 



## Indexación y Rebanadas

Al igual que las listas, es posible acceder a un elemento (indexación) o a un grupo de elementos (rebanadas) de en un arreglo.

<div align="center">
    <img src="https://github.com/MCD-ESPOL/PCD/blob/main/notebooks/images/arr_index_slicing.png?raw=true" width="50%">
</div>

In [ ]:
# Muestre los 3 primeros elementos de cada uno de los arreglos nombres_librosArr, genero_literarioArr y ventasArr


## Búsqueda (filtrado) de valores

A diferencia de las listas, es posible crear un arreglo (mascaraArr) a partir de aplicar una **condición** a todos los elementos de un arreglo.



<div align="center">
    <img src="https://github.com/MCD-ESPOL/PCD/blob/main/notebooks/images/arr_mask.png?raw=true" width="60%">
</div>

In [ ]:
# Cree una máscara con el arreglo genero_literarioArr cuyos elementos sean igual que 'Fantasía'


In [ ]:
# Utilice la máscara para filtrar el arreglo nombres_librosArr
# Muestre el arreglo resultante


In [ ]:
# Utilice la máscara para filtrar el arreglo ventasArr
# Muestre el arreglo resultante
# Obtenga y muestre el total 



## Archivos

Descargue los archivos [titles.csv](https://github.com/MCD-ESPOL/PCD/blob/main/notebooks/titles.csv) y [rating.csv](https://github.com/MCD-ESPOL/PCD/blob/main/notebooks/rating.csv). Luego, coloque los archivos csv en el mismo directorio que este cuaderno.

## Carga de datos


Utilice la función [loadtxt](https://numpy.org/doc/stable/reference/generated/numpy.loadtxt.html) para cargar los datos en en archivos.



<div align="center">
    <img src="https://github.com/MCD-ESPOL/PCD/blob/main/notebooks/images/arr_load1.png?raw=true" width="50%">
</div>




In [69]:
# Cargue los datos de 'titles.csv', cuyo delimitador es ';' y el tipo de dato (dtype) es str


In [71]:
# Cargue los datos de 'rating.csv', cuyo delimitador es ';' y el tipo de dato (dtype) es float


In [74]:
titlesArr, ratingArr

(array(['The Shawshank Redemption', 'The Godfather',
        'The Godfather: Part II', 'The Dark Knight', '12 Angry Men',
        "Schindler's List",
        'The Lord of the Rings: The Return of the King', 'Pulp Fiction',
        '"The Good, the Bad and the Ugly"',
        'The Lord of the Rings: The Fellowship of the Ring', 'Fight Club',
        'Forrest Gump', 'Inception',
        'The Lord of the Rings: The Two Towers', 'The Empire Strikes Back',
        'The Matrix', 'GoodFellas', "One Flew Over the Cuckoo's Nest",
        'Seven Samurai', 'Se7en', 'Life Is Beautiful', 'City of God',
        'The Silence of the Lambs', "It's a Wonderful Life", 'Star Wars',
        'Saving Private Ryan', 'The Green Mile', 'Spirited Away',
        'Interstellar', 'Parasite', 'Léon: The Professional', 'Harakiri',
        'The Usual Suspects', 'The Lion King', 'The Pianist',
        'Terminator 2: Judgment Day', 'Back to the Future',
        'American History X', 'Modern Times', 'Gladiator', 'Psycho'

In [76]:
np.unique(ratingArr)

array([8.1, 8.2, 8.3, 8.4, 8.5, 8.6, 8.7, 8.8, 8.9, 9. , 9.2, 9.3])

In [92]:
np.where((ratingArr >= 8.2) & (ratingArr <= 8.4)) 

(array([ 56,  57,  58,  59,  61,  62,  64,  65,  66,  67,  68,  69,  70,
         71,  72,  73,  74,  75,  76,  77,  78,  79,  80,  81,  82,  83,
         84,  85,  86,  87,  88,  89,  90,  91,  92,  93,  94,  95,  97,
         98,  99, 100, 101, 102, 103, 104, 105, 106, 107, 108, 109, 110,
        111, 112, 113, 114, 115, 116, 117, 118, 119, 120, 121, 122, 123,
        124, 125, 126, 127, 128, 129, 130, 131, 132, 133, 134, 135, 136,
        137, 138, 139, 140, 141, 143, 144, 145, 146, 147, 148, 149, 150,
        152, 153, 154, 156, 157, 159, 160, 161, 165, 167, 170, 173, 174,
        176, 179, 180, 181, 193, 196, 197, 200, 202, 203, 225, 229, 230,
        235, 238, 239, 244, 248], dtype=int64),)

## Ejercicios - I

1. Muestre el o los títulos de películas con la mayor calificación.
2. Muestre el o los títulos de películas con la menor calificación.
4. Solicite el ingreso de un valor de calificación y muestre las películas con dicha calificación.


## Ejercicios - II

1. Muestre la cantidad de películas cuya calificación sea mayor que 8.5
2. Muestre los nombres de las películas cuya calificación sea mayor que 8.5
3. Muestre la cantidad de películas cuya calificación esté en el rango \[8.2, 8.4\].

## Ejercicios -  III
1. Muestre los valores de calificaciones sin repetir. Utilice la función [unique](https://numpy.org/doc/stable/reference/generated/numpy.unique.html) para obtener un arreglo con valores únicos.
2. Muestre la cantidad de películas por cada calificación sin repetir.
3. Grafique los valores de calificaciones sin repetir _versus_ la cantidad de películas por calificaciones sin repetir.